### PR TITLE
Issues/728 cwl testing

### LINKED ISSUE #728: CWL toil workflow input file name collision
If two inputs to a cwl workflow step have the same name there will be a collision, for example:
```
2016-03-30 23:49:00,930 WARNING:toil.leader: f/z/jobmQtiTU:	---TOIL WORKER OUTPUT LOG---
2016-03-30 23:49:00,930 WARNING:toil.leader: f/z/jobmQtiTU:	INFO:rdflib:RDFLib Version: 4.2.1
2016-03-30 23:49:00,930 WARNING:toil.leader: f/z/jobmQtiTU:	INFO:toil.jobStores.fileJobStore:Jobstore directory is: /Users/ellrott/workspaces/mc3_cwl/test_workflow/test_store
2016-03-30 23:49:00,930 WARNING:toil.leader: f/z/jobmQtiTU:	Traceback (most recent call last):
2016-03-30 23:49:00,930 WARNING:toil.leader: f/z/jobmQtiTU:	  File "/Users/ellrott/workspaces/mc3_cwl/venv/lib/python2.7/site-packages/toil-3.2.0a2-py2.7.egg/toil/worker.py", line 283, in main
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	    fileStore=fileStore)
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	  File "/Users/ellrott/workspaces/mc3_cwl/venv/lib/python2.7/site-packages/toil-3.2.0a2-py2.7.egg/toil/job.py", line 1298, in _execute
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	    returnValues = self.run(fileStore)
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	  File "/Users/ellrott/workspaces/mc3_cwl/venv/lib/python2.7/site-packages/toil-3.2.0a2-py2.7.egg/toil/cwl/cwltoil.py", line 213, in run
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	    adjustFiles(cwljob, lambda x: getFile(fileStore, inpdir, *x, index=index))
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	  File "/Users/ellrott/workspaces/mc3_cwl/venv/lib/python2.7/site-packages/cwltool/process.py", line 139, in adjustFiles
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	    adjustFiles(rec[d], op)
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	  File "/Users/ellrott/workspaces/mc3_cwl/venv/lib/python2.7/site-packages/cwltool/process.py", line 137, in adjustFiles
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	    rec["path"] = op(rec["path"])
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	  File "/Users/ellrott/workspaces/mc3_cwl/venv/lib/python2.7/site-packages/toil-3.2.0a2-py2.7.egg/toil/cwl/cwltoil.py", line 213, in <lambda>
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	    adjustFiles(cwljob, lambda x: getFile(fileStore, inpdir, *x, index=index))
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	  File "/Users/ellrott/workspaces/mc3_cwl/venv/lib/python2.7/site-packages/toil-3.2.0a2-py2.7.egg/toil/cwl/cwltoil.py", line 123, in getFile
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	    raise Exception("Conflicting filesStoreID %s and %s both trying to link to %s" % (index.get(dstPath, None), fileStoreID, dstPath))
2016-03-30 23:49:00,931 WARNING:toil.leader: f/z/jobmQtiTU:	Exception: Conflicting filesStoreID n/1/tmpxw47b3.tmp and L/C/tmpCnIrNm.tmp both trying to link to /Users/ellrott/workspaces/mc3_cwl/test_workflow/tmpzJ7dNP/localTempDir/tn7MTGE/inp/out.txt
```

Example Workflow (works in cwltool, fails in cwltoil):
```
- class: CommandLineTool
  id: "#echo"
  inputs:
    - id: text
      type: string
      inputBinding: {}

  outputs:
    - id: "fileout"
      type: File
      outputBinding:
        glob: out.txt

  baseCommand: echo
  stdout: out.txt

- class: CommandLineTool
  id: "#cat"
  inputs:
    - id: file1
      type: File
      inputBinding:
        position: 1
    - id: file2
      type: File
      inputBinding:
        position: 2

  outputs:
    - id: "fileout"
      type: File
      outputBinding:
        glob: out.txt

  baseCommand: cat
  stdout: out.txt

- class: Workflow
  id: "#collision"

  inputs:
    - id: "#input_1"
      type: string
      
    - id: "#input_2"
      type: string
  
  outputs:
    - id: "#fileout"
      type: File
      source: "#cat_step.fileout"

  steps:
    - id: "#echo_1"
      run: "#echo"
      inputs:
        - id: "#echo_1.text"
          source: "#input_1"
      outputs:
        - id: "#echo_1.fileout"
    
    - id: "#echo_2"
      run: 
        import: "#echo"
      inputs:
        - id: "#collision.echo_2.text"
          source: "#input_2"
      outputs:
        - id: "#echo_2.fileout"
    
    - id: "#cat_step"
      run: 
        import: "#cat"
      inputs:
        - id: file1
          source: "#echo_1.fileout"
        - id: file2
          source: "#echo_2.fileout"
      outputs:
        - id: "#cat_step.fileout"
```

Sample input:
```
{
  "input_1" : "Hello World",
  "input_2" : "How are you?"
}
```

### PR DESCRIPTION
**Pull request by tetron**
Fixes #728, fixes some additional conformance tests errors, and adds CWL conformance testing to toil testing infrastructure.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "src/toil/test/cwl/spec"]
+	path = src/toil/test/cwl/spec
+	url = https://github.com/common-workflow-language/common-workflow-language.git
```
Comment (hannes-ucsc) [inline .gitmodules]: Shouldn't this specify a tag/commit?

**Comment by hannes-ucsc** [comment]
Submodules are kinda intrusive on the development process so I need to be convinced that a submodule is the only way to pull in the spec. Is the submodule only needed to for the tests? Can we clone it programmatically as part of the test setup?

**Comment by tetron** [comment]
Yes, I wasn't sure how you felt in terms of being able to clone everything up front vs going out to the Internet during the test run (which can result in annoying false positives when there are network hiccups).  I will remove the submodule and have the test clone the repo on the fly.

**Comment by hannes-ucsc** [comment]
Thank you.

**Comment by tetron** [comment]
Done.  Rebased master and squashed commits.

File: src/toil/test/cwl/cwlTest.py [hunk 2]
```diff
@@ -14,30 +14,47 @@
...
                 u'checksum': u'sha1$b9214658cc453331b62c2282b772a5c063dbd284'}})
+
+
+    def test_run_conformance(self):
+        rootDir = self._projectRootPath()
+        cwlSpec = os.path.join(rootDir, 'src/toil/test/cwl/spec')
+        if os.path.exists(cwlSpec):
+            subprocess.check_call(["git", "pull"], cwd=cwlSpec)
+        else:
+            subprocess.check_call(["git", "clone", "https://github.com/common-workflow-language/common-workflow-language.git", cwlSpec])
```
Comment (hannes-ucsc) [inline src/toil/test/cwl/cwlTest.py]: So you always want the latest commit on the default branch of the repo? If we pin the cwltool dependency setup.py in shouldn't we also pin this?

**Comment by tetron** [comment]
Pinned common-workflow-language repo version.  Also bumped cwltool pin for dependency on a function imported from most recent cwltool.

**Comment by tetron** [comment]
I don't know why this is marked as "checks failed" with 0 failing tests?

**Comment by hannes-ucsc** [comment]
It's the check for dirty files that is failing. I added this to "encourage" our developers from letting their unit tests leave garbage around in the source tree.

Can you try adding `/src/toil/test/cwl/spec` to .gitignore? If that doesn't fix it I will take care of it.

File: src/toil/test/cwl/cwlTest.py [hunk 3]
```diff
@@ -14,30 +14,48 @@
...
+
+    def test_run_conformance(self):
+        rootDir = self._projectRootPath()
+        cwlSpec = os.path.join(rootDir, 'src/toil/test/cwl/spec')
+        if os.path.exists(cwlSpec):
+            subprocess.check_call(["git", "fetch"], cwd=cwlSpec)
+        else:
+            subprocess.check_call(["git", "clone", "https://github.com/common-workflow-language/common-workflow-language.git", cwlSpec])
+        subprocess.check_call(["git", "checkout", "4daac6db94ccf626c4509fe0c3e645ceda23f50a"], cwd=cwlSpec)
+        subprocess.check_call(["./run_test.sh", "RUNNER=cwltoil", "DRAFT=draft-3"], cwd=os.path.join(os.getcwd(), "src/toil/test/cwl/spec"))
```
Comment (hannes-ucsc) [inline src/toil/test/cwl/cwlTest.py]: Can os.getcwd() be replaced with `rootDir`?

**Comment by tetron** [comment]
Done and done, let's see how Jenkins likes it.

**Comment by tetron** [comment]
Tests passing now, good to merge?

**Comment by hannes-ucsc** [comment]
The commits should probably have been squashed and there is a newline missing at the end but I didn't want to do another roundtrip.

Thank you.

**Pull request merged**